vortex-layout/src/layouts/collect.rs

```diff
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright the Vortex contributors
+
+use std::sync::Arc;
+
+use async_stream::try_stream;
+use async_trait::async_trait;
+use futures::{StreamExt, pin_mut};
+use vortex_array::ArrayContext;
+use vortex_array::arrays::ChunkedArray;
+use vortex_error::{VortexExpect, VortexResult};
+use vortex_io::runtime::Handle;
+
+use crate::segments::SegmentSinkRef;
+use crate::sequence::{
+    SendableSequentialStream, SequencePointer, SequentialStream, SequentialStreamAdapter,
+};
+use crate::{LayoutRef, LayoutStrategy};
+
+/// A strategy that collects all chunks and turns them into a single array chunk to pass into
+/// a child strategy.
+pub struct CollectStrategy {
+    child: Arc<dyn LayoutStrategy>,
+}
+
+impl CollectStrategy {
+    pub fn new<S: LayoutStrategy>(child: S) -> CollectStrategy {
+        CollectStrategy {
+            child: Arc::new(child),
+        }
+    }
+}
+
+#[async_trait]
+impl LayoutStrategy for CollectStrategy {
+    async fn write_stream(
+        &self,
+        ctx: ArrayContext,
+        segment_sink: SegmentSinkRef,
+        stream: SendableSequentialStream,
+        eof: SequencePointer,
+        handle: Handle,
+    ) -> VortexResult<LayoutRef> {
+        // Read the whole stream, then write one Chunked stream to the inner thing
+        let dtype = stream.dtype().clone();
+
+        let _dtype = dtype.clone();
+        let collected_stream = try_stream! {
+            pin_mut!(stream);
+
+            let mut chunks = Vec::new();
+            let mut latest_sequence_id = None;
+            while let Some(chunk) = stream.next().await {
+                let (sequence_id, chunk) = chunk?;
+                latest_sequence_id = Some(sequence_id);
+                chunks.push(chunk);
+            }
+
+            let collected = ChunkedArray::try_new(chunks, _dtype)?.to_array();
```

Comment:
maybe we canonicalize here instead of writing a flat chunked array?

Reply:
In practice that ends up happening b/c the `child` here is the CompressStrategy, so it will canonicalize + compress internally.

In general I tried to mirror this on the RepartitionStrategy which yields ChunkedArray as well. I think the LayoutStrategies are a bit weird tbh and there are a few unexpected things like this.

Reply:
I thought repartition strategy will have a chunked strategy as a child so it will flatten the chunked array, in this case I think we go straight to flat.